[JSC] Improve BBQ's rotate and shift implementations

mikhailramalho opened the pull request:
#### b5819712d9f5b5ef54b649b04d5ad1aebb4aed01
<pre>
[JSC] Improve BBQ&apos;s rotate and shift implementations
<a href="https://bugs.webkit.org/show_bug.cgi?id=302406">https://bugs.webkit.org/show_bug.cgi?id=302406</a>

Reviewed by Yusuke Suzuki.

This PR is mostly focused on improving the rotate and shift implementations for
ARMv7, but also includes small thcanges to 64-bit arm which I&apos;ll describe
below.

For 32-bit:
- Implemented rotateLeft64, rotateRight64, lshift64, urshift64, rshift64:
  similar to some fp functions (e.g., truncateDoubleToUint64) we need to pass 2
  scratches to implement the algorithms. One benefit of moving the shift
  operations to the backend is that we have an extra scratch (r12), so things are
  slightly easier to implement.
- Removed lshiftUnchecked(), rshiftUnchecked(), urshiftUnchecked() since we can
  access the instructions directly.
- Implemented several optimizations for shift and rotate: we see a .5%
  reduction in code size in JetStream3&apos;s tfjs-wasm.js.
- The rotate algorithm should now be safe to run when registers overlap, which
  was causing a crash in JetStream3&apos;s tfjs-wasm.js.

For 64-bit arm:
- Implemented rotateLeft32 and rotateLeft64: this allows us to unify both
  x86_64 and ARM64 code paths.

* Source/JavaScriptCore/assembler/MacroAssemblerARM64.h:
(JSC::MacroAssemblerARM64::rotateLeft32):
(JSC::MacroAssemblerARM64::rotateLeft64):
* Source/JavaScriptCore/assembler/MacroAssemblerARMv7.h:
(JSC::MacroAssemblerARMv7::rotateLeft64):
(JSC::MacroAssemblerARMv7::rotateRight64):
(JSC::MacroAssemblerARMv7::lshift64):
(JSC::MacroAssemblerARMv7::urshift64):
(JSC::MacroAssemblerARMv7::rshift64):
(JSC::MacroAssemblerARMv7::lshiftUnchecked): Deleted.
(JSC::MacroAssemblerARMv7::rshiftUnchecked): Deleted.
(JSC::MacroAssemblerARMv7::urshiftUnchecked): Deleted.
* Source/JavaScriptCore/wasm/WasmBBQJIT.cpp:
(JSC::Wasm::BBQJITImpl::BBQJIT::addI32Rotl):
* Source/JavaScriptCore/wasm/WasmBBQJIT.h:
* Source/JavaScriptCore/wasm/WasmBBQJIT32_64.cpp:
(JSC::Wasm::BBQJITImpl::BBQJIT::addI64Shl):
(JSC::Wasm::BBQJITImpl::BBQJIT::addI64ShrS):
(JSC::Wasm::BBQJITImpl::BBQJIT::addI64ShrU):
(JSC::Wasm::BBQJITImpl::BBQJIT::addI64Rotl):
(JSC::Wasm::BBQJITImpl::BBQJIT::addI64Rotr):
(JSC::Wasm::BBQJITImpl::BBQJIT::rotI64Helper): Deleted.
* Source/JavaScriptCore/wasm/WasmBBQJIT64.cpp:
(JSC::Wasm::BBQJITImpl::BBQJIT::addI64Rotl):

Canonical link: <a href="https://commits.webkit.org/303502@main">https://commits.webkit.org/303502@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2280930fff24fc63b45c574ce89dd996cbf3eed2

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/130861 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/3184 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/41820 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/138289 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/82520 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/3166 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/3027 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/99707 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/67516 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/133807 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/2281 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/117177 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/80401 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/2201 "Passed tests") | [⏳ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/macOS-Tahoe-Debug-API-Tests-EWS "Waiting to run tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/81544 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/122877 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/110794 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/35306 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/140767 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/129314 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/2928 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/35809 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/108222 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/2974 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/2640 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/108142 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/27863 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/2244 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/113508 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/55962 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/2996 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/31935 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/162331 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/2817 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/66388 "Built successfully") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/40494 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/3017 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/2925 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

mikhailramalho commented on Source/JavaScriptCore/assembler/MacroAssemblerARMv7.h:
```diff
@@ -354,6 +354,343 @@ class MacroAssemblerARMv7 : public AbstractMacroAssembler<Assembler> {
+            m_assembler.mov(destLo, scratch0);
+            m_assembler.mov(destHi, scratch1);
+            return;
+        }
```
Just realized this should be done before the bit 5 test, so we can save two moves. Same for the others

Constellation commented on Source/JavaScriptCore/assembler/MacroAssemblerARM64.h:
```diff
@@ -1425,6 +1444,25 @@ class MacroAssemblerARM64 : public AbstractMacroAssembler<Assembler> {
+    void rotateLeft64(RegisterID src, RegisterID shiftAmount, RegisterID dest)
+    {
+        RegisterID scratch = getCachedDataTempRegisterIDAndInvalidate();
+        neg32(shiftAmount, scratch);
```
This should be neg64, right?

mikhailramalho replied:
Yep, sorry. Fixed.

Constellation commented on Source/JavaScriptCore/assembler/MacroAssemblerARM64.h:
```diff
@@ -1425,6 +1444,25 @@ class MacroAssemblerARM64 : public AbstractMacroAssembler<Assembler> {
+    {
+        if (!imm.m_value) [[unlikely]]
+            return move(src, dest);
+        rotateRight64(src, TrustedImm32(-imm.m_value), dest);
```
What happens if this imm.m_value is INT32_MIN?

mikhailramalho replied:
That would be ub, right? I changed it to `rotateRight64(src, TrustedImm32(-(imm.m_value & 63)), dest);`

Constellation commented on Source/JavaScriptCore/assembler/MacroAssemblerARM64.h:
```diff
@@ -1408,6 +1408,25 @@ class MacroAssemblerARM64 : public AbstractMacroAssembler<Assembler> {
+    {
+        if (!imm.m_value) [[unlikely]]
+            return move(src, dest);
+        rotateRight32(src, TrustedImm32(-imm.m_value), dest);
```
Ditto.

mikhailramalho replied:
Changed to `rotateRight32(src, TrustedImm32(-(imm.m_value & 31)), dest);`

justinmichaud commented on Source/JavaScriptCore/assembler/MacroAssemblerARM64.h:
```diff
@@ -1408,6 +1408,25 @@ class MacroAssemblerARM64 : public AbstractMacroAssembler<Assembler> {
+
+    void rotateLeft32(RegisterID src, RegisterID shiftAmount, RegisterID dest)
+    {
+        RegisterID scratch = getCachedDataTempRegisterIDAndInvalidate();
```
I wonder if we should take in a tmp instead. If we ever want to use this in air, I believe that would be required

Constellation replied:
Yes. But if it is not, I think we do not need to take scratch. And if it is like this, it should be neg and rotate in Air instead of using this

mikhailramalho committed:
[JSC] Improve BBQ's rotate and shift implementations
https://bugs.webkit.org/show_bug.cgi?id=302406

Reviewed by Yusuke Suzuki.

This PR is mostly focused on improving the rotate and shift implementations for
ARMv7, but also includes small thcanges to 64-bit arm which I'll describe
below.

For 32-bit:
- Implemented rotateLeft64, rotateRight64, lshift64, urshift64, rshift64:
  similar to some fp functions (e.g., truncateDoubleToUint64) we need to pass 2
  scratches to implement the algorithms. One benefit of moving the shift
  operations to the backend is that we have an extra scratch (r12), so things are
  slightly easier to implement.
- Removed lshiftUnchecked(), rshiftUnchecked(), urshiftUnchecked() since we can
  access the instructions directly.
- Implemented several optimizations for shift and rotate: we see a .5%
  reduction in code size in JetStream3's tfjs-wasm.js.
- The rotate algorithm should now be safe to run when registers overlap, which
  was causing a crash in JetStream3's tfjs-wasm.js.

For 64-bit arm:
- Implemented rotateLeft32 and rotateLeft64: this allows us to unify both
  x86_64 and ARM64 code paths.

* Source/JavaScriptCore/assembler/MacroAssemblerARM64.h:
(JSC::MacroAssemblerARM64::rotateLeft32):
(JSC::MacroAssemblerARM64::rotateLeft64):
* Source/JavaScriptCore/assembler/MacroAssemblerARMv7.h:
(JSC::MacroAssemblerARMv7::rotateLeft64):
(JSC::MacroAssemblerARMv7::rotateRight64):
(JSC::MacroAssemblerARMv7::lshift64):
(JSC::MacroAssemblerARMv7::urshift64):
(JSC::MacroAssemblerARMv7::rshift64):
(JSC::MacroAssemblerARMv7::lshiftUnchecked): Deleted.
(JSC::MacroAssemblerARMv7::rshiftUnchecked): Deleted.
(JSC::MacroAssemblerARMv7::urshiftUnchecked): Deleted.
* Source/JavaScriptCore/wasm/WasmBBQJIT.cpp:
(JSC::Wasm::BBQJITImpl::BBQJIT::addI32Rotl):
* Source/JavaScriptCore/wasm/WasmBBQJIT.h:
* Source/JavaScriptCore/wasm/WasmBBQJIT32_64.cpp:
(JSC::Wasm::BBQJITImpl::BBQJIT::addI64Shl):
(JSC::Wasm::BBQJITImpl::BBQJIT::addI64ShrS):
(JSC::Wasm::BBQJITImpl::BBQJIT::addI64ShrU):
(JSC::Wasm::BBQJITImpl::BBQJIT::addI64Rotl):
(JSC::Wasm::BBQJITImpl::BBQJIT::addI64Rotr):
(JSC::Wasm::BBQJITImpl::BBQJIT::rotI64Helper): Deleted.
* Source/JavaScriptCore/wasm/WasmBBQJIT64.cpp:
(JSC::Wasm::BBQJITImpl::BBQJIT::addI64Rotl):

Canonical link: https://commits.webkit.org/303502@main
